lib/Sema/CSDiagnostics.cpp

```diff
@@ -1928,6 +1929,21 @@ bool ContextualFailure::diagnoseAsError() {
     if (diagnoseYieldByReferenceMismatch())
       return true;
 
+    if (CTP == CTP_ForEachStmt) {
+      if (FromType->is<ProtocolType>()) {
+        emitDiagnostic(anchor->getLoc(), diag::type_cannot_conform,
+                       /*isExistentialType=*/false, FromType, ToType);
```

Comment:
Should `isExistentialType` here be true? I think that's why we're getting the general `type 'Sequence' cannot conform to 'Sequence'` instead of the special protocol message in the last test case

Reply:
I actually want to keep it reporting `for-in ...` diagnostic for existentials, I added this special case for examples like this - `func foo(_ s: Sequence) { for x in s { .. } }`  where `s` is incorrect use of protocol type, otherwise it would produce `for-in loop requires 'Sequence' to conform to 'Sequence'` which doesn't really make sense.

Reply:
I was thinking about extending it to more types e.g.  all that can't possibly conform to protocols but I don't like that it's unclear where that requirement is coming from...

Reply:
I understand how the special `for-in` diagnostic can be confusing here, I just mean the `isExistentialType` argument to the `type_cannot_conform` diagnostic in this special case, so it reads `value of protocol type Sequence cannot conform...` rather than `type Sequence cannot conform...`

Reply:
But in that case it's not a value of a protocol type though...

Reply:
In the case of `func foo(_ s: Sequence) { for x in s { .. } }`, `s` has type `Sequence`, and `Sequence` is a protocol. Besides the fact that using `Sequence` as an existential type is not actually allowed due to associated type requirements, what am I missing?

Reply:
You are right, I can just check for `isExistentialType()` instead. For some reason I just thought that in this situation it's not actually going to be a value of protocol type but the type itself...

Reply:
Fixed!